pkg/sdn/plugin/multitenant.go

```diff
@@ -0,0 +1,126 @@
+package plugin
+
+import (
+	"fmt"
+
+	log "github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/container"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+	utilwait "k8s.io/kubernetes/pkg/util/wait"
+)
+
+func (node *OsdnNode) MultitenantStartNode() error {
+	if err := node.VnidStartNode(); err != nil {
+		return err
+	}
+	if err := node.SetupEgressNetworkPolicy(); err != nil {
+		return err
+	}
+
+	go utilwait.Forever(node.watchServices, 0)
+	return nil
+}
+
+func (node *OsdnNode) updatePodNetwork(namespace string, oldNetID, netID uint32) error {
+	// FIXME: this is racy; traffic coming from the pods gets switched to the new
+	// VNID before the service and firewall rules are updated to match. We need
+	// to do the updates as a single transaction (ovs-ofctl --bundle).
+
+	pods, err := node.GetLocalPods(namespace)
+	if err != nil {
+		return err
+	}
+	services, err := node.kClient.Services(namespace).List(kapi.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	errList := []error{}
+
+	// Update OF rules for the existing/old pods in the namespace
+	for _, pod := range pods {
+		err = node.UpdatePod(pod.Namespace, pod.Name, kubetypes.DockerID(getPodContainerID(&pod)))
+		if err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	// Update OF rules for the old services in the namespace
+	for _, svc := range services.Items {
+		if !kapi.IsServiceIPSet(&svc) {
+			continue
+		}
+
+		if err = node.DeleteServiceRules(&svc); err != nil {
+			log.Error(err)
+		}
+		if err = node.AddServiceRules(&svc, netID); err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	// Update namespace references in egress firewall rules
+	if err = node.UpdateEgressNetworkPolicyVNID(namespace, oldNetID, netID); err != nil {
+		errList = append(errList, err)
+	}
+
+	return kerrors.NewAggregate(errList)
+}
+
+func isServiceChanged(oldsvc, newsvc *kapi.Service) bool {
+	if len(oldsvc.Spec.Ports) == len(newsvc.Spec.Ports) {
+		for i := range oldsvc.Spec.Ports {
+			if oldsvc.Spec.Ports[i].Protocol != newsvc.Spec.Ports[i].Protocol ||
+				oldsvc.Spec.Ports[i].Port != newsvc.Spec.Ports[i].Port {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+func (node *OsdnNode) watchServices() {
+	services := make(map[string]*kapi.Service)
+	RunEventQueue(node.kClient, Services, func(delta cache.Delta) error {
```

Comment:
Should probably update this to use eventqueue.NewEventQueueForStore() and make 'services' a cache.Store, but that could be done after.